File: vtal/custom-template-utils/src/java/src/com/cisco/nso/common/ctutils/CustomTemplateUtils.java
/*
 ***************************************************
 * Copyright (c) 2020 Cisco Systems, Inc.
 *
 * All rights reserved.
 ***************************************************
 */
package com.cisco.nso.common.ctutils;

import java.util.Map;
import com.tailf.conf.ConfException;
import com.tailf.dp.services.ServiceContext;
import com.tailf.navu.NavuList;
import com.tailf.navu.NavuNode;

/**
 *
 * @author abhatta
 *
 *         The interface which provides the api methods, to integrate with core function packs.
 *
 */
public interface CustomTemplateUtils {
  /**
   * Apply multiple custom templates to a device with extra variable map.
   *
   * @param context {@link ServiceContext} Mandatory parameter must be passed from where this method
   *        is called.
   * @param deviceName {@link String} Mandatory parameter name of the device to which custom
   *        templates will be applied.
   * @param keyPath {@link String} Mandatory parameter keyPath till {@link NavuList} custom-template
   *        in yang model.
   * @param extraVariables {@link Map} of variable name value pair, which will be used for all
   *        iterations. Pass {@link null}, if there is no extra variables to be used while applying
   *        templates.User provided template variables will overwrite the variables present in this
   *        map if variable name matches.
   * @return false if apply-custom-template flag is false, else true.
   * @throws ConfException
   */
  static boolean applyCustomTemplates(ServiceContext context, String deviceName, String keyPath,
      Map<String, String> extraVariables) throws ConfException {
    return CTUtilsFunction.applyCustomTemplates(context, deviceName, keyPath, extraVariables);
  }

  /**
   * Apply a single custom template to a device with extra variable map.
   *
   * @param context {@link ServiceContext} Mandatory parameter must be passed from where this method
   *        is called.
   * @param deviceName {@link String} Mandatory parameter name of the device to which custom
   *        template will be applied.
   * @param keyPath {@link String} Mandatory parameter keyPath till {@link NavuNode}
   *        custom-template{%s} in yang model.
   * @param extraVariables {@link Map} of variable name value pair, which will be used for all
   *        iterations. Pass {@link null}, if there is no extra variables to be used while applying
   *        templates.User provided template variables will overwrite the variables present in this
   *        map if variable name matches.
   * @return false if apply-custom-template flag is false, else true.
   * @throws ConfException
   */
  static boolean applyCustomTemplate(ServiceContext context, String deviceName, String keyPath,
      Map<String, String> extraVariables) throws ConfException {
    return CTUtilsFunction.applyCustomTemplate(context, deviceName, keyPath, extraVariables);
  }

  /**
   * Apply any template to a device with provided variable map only one time.
   *
   * @param context {@link ServiceContext} Mandatory parameter must be passed from where this method
   *        is called.
   * @param deviceName {@link String} Mandatory parameter name of the device to which custom
   *        templates are applied.
   * @param templateName {@link String} Mandatory parameter name of the template to be applied.
   * @param variables {@link Map} of variable name value pair to be used while applying the
   *        template.
   * @return false if apply-custom-template flag is false, else true.
   * @throws ConfException
   */
  static boolean applyTemplate(ServiceContext context, String deviceName, String templateName,
      Map<String, String> variables) throws ConfException {
    return CTUtilsFunction.applyTemplate(context, deviceName, templateName, variables);
  }
}

File: vtal/custom-template-utils/python/custom_template_utils/custom_template_utils.py
from . import utils
from . import ct_utils


def apply_custom_template(self, root, service, device_name, ct_node, extra_vars=None):
    '''
    Apply a single custom-template represented by ct-node on a device.
        self: ServiceCallbacks instance
        root: root node
        service: service node and the custom-template will be applied on this context.
        device-name: Name of the device on which the custom-template will be applied.
        ct_node: cutom_template node
        extra_vars: Dictionary {variable_name:variable_value} of extra variables.
            These variables will be over-written by the custom-template
            variables if variable name matches.
    '''
    if utils.validate_input(root, service, device_name, ct_node=ct_node):
        self.log.info('====================================================================')
        self.log.info('custom-template is triggered for ', ct_node._path)
        ct_utils.process_ct_node(self, root, service, device_name, ct_node, extra_vars)
        return True
    else:
        self.log.info('apply-custom-template flag is false.')
        return False


def apply_custom_templates(self, root, service, device_name, ct_list_node, extra_vars=None):
    '''
    Apply multiple custom-templates represented by ct_list_node on a device.
        self: ServiceCallbacks instance
        root: root node
        service: service node and all the custom-templates will be applied on this context.
        device-name: Name of the device on which all the custom-templates will be applied.
        ct_list_node: cutom_template list node
        extra_vars: Dictionary {variable_name:variable_value} of extra variables.
            These variables will be over-written by the custom-template
            variables if variable name matches.
    '''
    if utils.validate_input(root, service, device_name, ct_node=ct_list_node):
        self.log.info('====================================================================')
        self.log.info('custom-templates are triggered for ', ct_list_node._path)
        for ct_node in ct_list_node:
            ct_utils.process_ct_node(self, root, service, device_name, ct_node, extra_vars)
        return True
    else:
        self.log.info('apply-custom-template flag is false.')
        return False


def apply_template(self, root, service, device_name, template_name, variables=None):
    '''
    Apply any template on a device once with provided variables.
        self: ServiceCallbacks instance
        root: root node
        service: service node and the custom-template will be applied on this context.
        device-name: Name of the device on which the custom-template will be applied.
        template_name: Name of the cutom_template which will be applied.
        variables: Dictionary {variable_name:variable_value} of variables in the custom-template.
    '''
    if utils.validate_input(root, service, device_name, template_name=template_name):
        self.log.info('====================================================================')
        self.log.info('Applying custom-template: ', template_name, ' once on device: ', device_name)
        ct_utils.process_ct(self, root, service, device_name, template_name, variables)
        return True
    else:
        self.log.info('apply-custom-template flag is false.')
        return False
